### ISSUE TITLE
`unconditional_panic` array bounds check doesn't work through `Deref`

### ISSUE THREAD

**Issue by e2-71828** [rust-lang/rust#109260]
(Initially reported on [URLO](https://users.rust-lang.org/t/im-really-surprised-that-this-compiles/90794))

In the following program:

```rust
fn main() {
    let mut arr_0: [u32; 10] = [0, 1, 2, 3, 4, 5, 6, 7, 8, 9];

    let ptr_0 = Box::new(arr_0);

    let ptr_v = ptr_0[11];
    let arr_v = arr_0[11];
    println!("v:{}", ptr_v);
}
```

The out-of-bounds access of `ptr_0[11]` is not flagged at compile time, but the similar access of `arr_0[11]` is.

**Comment by clubby789** [comment]
It looks like the MIR ends up as 
```rs
_7 = const false;
assert(move _7, "index out of bounds: the length is {} but the index is {}", move _6, _5) -> [success: bb2, unwind: bb4];
```
and the lint isn't able to look through that constant to identify it

**Comment by veera-sivarajan** [comment]
Fixed by #129517: https://rust.godbolt.org/z/YqKf4YhK3
